rabbit/rabbittransport.go

```diff
@@ -0,0 +1,426 @@
+	headers["Endpoint"] = req.Endpoint()
+
+	select {
+	case <-t.Ready():
+	case <-timeout.C:
+		log.Warnf("[Typhon:RabbitTransport] Timed out after %s waiting for ready", _timeout.String())
+		return nil, transport.ErrTimeout
+	}
+
+	if err := t.connection().Publish(Exchange, req.Service(), amqp.Publishing{
+		CorrelationId: req.Id(),
+		Timestamp:     time.Now().UTC(),
+		Body:          req.Payload(),
+		ReplyTo:       t.replyQueue,
+		Headers:       headersToTable(headers),
+	}); err != nil {
+		log.Errorf("[Typhon:RabbitTransport] Failed to publish: %v", err)
+		return nil, err
+	}
+
+	select {
+	case rsp := <-rspChan:
+		return rsp, nil
+	case <-timeout.C:
+		log.Warnf("[Typhon:RabbitTransport] Timed out after %s waiting for response to %s", _timeout.String(),
+			req.Id())
+		return nil, transport.ErrTimeout
+	}
+}
+
+func (t *rabbitTransport) listenReplies() error {
+	if err := t.connection().Channel.DeclareReplyQueue(t.replyQueue); err != nil {
+		log.Criticalf("[Typhon:RabbitTransport] Failed to declare reply queue %s: %v", t.replyQueue, err)
+		os.Exit(1)
+		return err
+	}
+
+	deliveries, err := t.connection().Channel.ConsumeQueue(t.replyQueue)
+	if err != nil {
+		log.Criticalf("[Typhon:RabbitTransport] Failed to consume from reply queue %s: %v", t.replyQueue, err)
+		os.Exit(1)
+		return err
+	}
+
+	log.Debugf("[Typhon:RabbitTransport] Listening for replies on %s…", t.replyQueue)
+	t.connM.RLock()
+	readyC := t.connReady
+	t.connM.RUnlock()
+	select {
+	case <-readyC:
+		// Make sure not to close the channel if it's already closed
+	default:
+		close(readyC)
+	}
+
+	for {
+		select {
+		case delivery, ok := <-deliveries:
+			if !ok {
+				log.Warnf("[Typhon:RabbitTransport] Delivery channel %s closed", t.replyQueue)
+				return ErrDeliveriesClosed
+			}
+			go t.handleRspDelivery(delivery)
+
+		case <-t.tomb.Dying():
+			log.Info("[Typhon:RabbitTransport] Reply listener terminating (tomb death)")
+			return tomb.ErrDying
+		}
+	}
+}
+
+func (t *rabbitTransport) deliveryToMessage(delivery amqp.Delivery, msg message.Message) {
+	msg.SetId(delivery.CorrelationId)
+	msg.SetHeaders(tableToHeaders(delivery.Headers))
+	msg.SetHeader("X-Rabbit-ReplyTo", delivery.ReplyTo)
+	msg.SetPayload(delivery.Body)
+	switch service := delivery.Headers["Service"].(type) {
+	case string:
+		msg.SetService(service)
+	}
+	switch endpoint := delivery.Headers["Endpoint"].(type) {
+	case string:
+		msg.SetEndpoint(endpoint)
+	}
+}
+
+func (t *rabbitTransport) handleReqDelivery(delivery amqp.Delivery, reqChan chan<- message.Request) {
+	logId := t.logId(delivery)
+	enc := delivery.Headers["Content-Encoding"].(string)
+	switch enc {
+	case "request":
+		req := message.NewRequest()
+		t.deliveryToMessage(delivery, req)
+
+		timeout := time.NewTimer(chanSendTimeout)
+		defer timeout.Stop()
+		select {
+		case reqChan <- req:
+		case <-timeout.C:
+			log.Errorf("[Typhon:RabbitTransport] Could not deliver request %s after %s: receiving channel is full",
```

Comment:
In the case of a timeout pushing into the request delivery channel we drop the message here? Is that correct? In which case the corresponding client which sent this request would wait for the full timeout at the other end, rather than receiving some form of out of capacity error? Again we can deal with this later :smile: